src/auspex/plotting.py

```diff
@@ -6,66 +6,78 @@
 #
 #    http://www.apache.org/licenses/LICENSE-2.0
 
-import threading
+from threading import Thread
 import subprocess
 import psutil
 import os
+import json
 import sys
 import tempfile
 import time
-
+import asyncio
+import zmq
+import zmq.asyncio
 from auspex.log import logger
 
-class BokehServerProcess(object):
-    def __init__(self, notebook=False):
-        super(BokehServerProcess, self).__init__()
-        self.run_in_notebook = notebook
-        self.pid_filename = os.path.join(tempfile.gettempdir(), "auspex_bokeh.pid")
+class MatplotServerThread(Thread):
 
-    def run(self):
-        # start a Bokeh server if one is not already running
-        pid = self.read_session_pid()
-        if pid:
-            self.p = psutil.Process(pid)
-            logger.info("Using existing Bokeh server")
-            return
-        logger.info("Starting Bokeh server")
-        args = ["bokeh", "serve", "--port", "5006", "--allow-websocket-origin=localhost:8888", "--allow-websocket-origin=localhost:8889", "--allow-websocket-origin=localhost:8890"]
-        self.p = subprocess.Popen(args, env=os.environ.copy())
-        self.write_session_pid()
-        # sleep to give the Bokeh server a chance to start
-        # TODO replace this with some bokeh client API call that
-        # verifies that the server is running
-        time.sleep(3)
+    def __init__(self, plot_desc={}, status_port = 7771, data_port = 7772):
+        super(MatplotServerThread, self).__init__()
+        self.plot_desc = plot_desc
+        self.status_port = status_port
+        self.data_port = data_port
+        self.daemon = True
+        self.stopped = False
+        self.start()
+
+    async def poll_sockets(self):
+        while not self.stopped:
+            evts = dict(await self.poller.poll(50))
+            if self.status_sock in evts and evts[self.status_sock] == zmq.POLLIN:
+                ident, msg = await self.status_sock.recv_multipart()
+                print("Got {} from {}".format(msg.decode(), ident.decode()))
+                if msg == b"WHATSUP":
+                    await self.status_sock.send_multipart([ident, b"HI!", json.dumps(self.plot_desc).encode('utf8')])
+            await asyncio.sleep(0.010)
+
+    async def _send(self, name, data, msg):
+        md = dict(
+            dtype = str(data.dtype),
+            shape = data.shape,
+        )
+        await self.data_sock.send_multipart([msg.encode(), name.encode(), json.dumps(md).encode(), data])
 
-    def terminate(self):
-        if self.p:
-            print("Killing bokeh server process {}".format(self.p.pid))
+    def send(self, name, data, msg="data"):
+        self._loop.create_task(self._send(name, data, msg=msg))
+
+    def stop(self):
+        self.stopped = True
+        pending = asyncio.Task.all_tasks(loop=self._loop)
+        self._loop.stop()
+        time.sleep(1)
```

Comment:
Oh, on the first read through I didn't catch the call to `self._loop.stop()` *before* canceling the pending tasks.

Another version of this that *might* work better is to catch `asyncio.CancelledError` after `self._loop.run_forever()` on line 79, and change that line to `self._loop.run_until_complete()`. Then I think you can just `cancel()` the tasks without having to stop and restart the event loop.